sys/auto_init/auto_init.c

```diff
@@ -23,284 +23,225 @@
 #include <stdio.h>
 #include "sched.h"
 #include "auto_init.h"
+#include "auto_init_macros.h"
 #include "kernel_defines.h"
-#include "log.h"
+
+XFA_INIT_CONST(auto_init_fn, auto_init_xfa);
+#if IS_USED(MODULE_AUTO_INIT_ZTIMER)
+extern void ztimer_init(void);
+AUTO_INIT_MODULE(ztimer, AUTO_INIT_PRIO_ZTIMER, ztimer_init);
+#endif
+#if IS_USED(MODULE_AUTO_INIT_XTIMER) && !IS_USED(MODULE_ZTIMER_XTIMER_COMPAT)
+extern void xtimer_init(void);
+AUTO_INIT_MODULE(xtimer, AUTO_INIT_PRIO_XTIMER, xtimer_init);
+#endif
+#if IS_USED(MODULE_AUTO_INIT_RANDOM)
+extern void auto_init_random(void);
+AUTO_INIT_MODULE(random, AUTO_INIT_PRIO_RANDOM, auto_init_random);
+#endif
```

Comment:
I doubt we need the full configurability of Kconfig and the redirection layers. If this just went in the direction of the below, there'd less complexity around.

```patch
+ #define PRIO_TIMERS(n) CONCAT(1000, n)
```

```suggestion
extern void ztimer_init(void);
AUTO_INIT_MODULE(ztimer, PRIO_TIMERS(00), ztimer_init);
#endif
#if IS_USED(MODULE_AUTO_INIT_XTIMER) && !IS_USED(MODULE_ZTIMER_XTIMER_COMPAT)
extern void xtimer_init(void);
AUTO_INIT_MODULE(xtimer, PRIO_TIMERS(01), xtimer_init);
#endif
#if IS_USED(MODULE_AUTO_INIT_RANDOM)
extern void auto_init_random(void);
AUTO_INIT_MODULE(random, PRIO_TIMERS(99), auto_init_random);
#endif
```

(hm, github wouldn't show this as a patch on its own; isn't supposed to be applied anyway, just food for thought / further discussion)

Reply:
With redirection layer you mean the reordering of auto-init modules? There would probably no reason to reorder modules if dependencies were modeled correctly. Maybe it was just an example but I would not express all dependencies relative to timers. For example random does not seem to depend on timers.

Reply:
I'm not talking about all dependencies relative to timers, more like having groups of known points (in Linux it'd be "root is mounted", "home directories are mounted", "network is up") of which timers are just the first in the examples (and random initialization should take input from real time, if there is any).

Macros like PRIO_TIMERS would merely serve to give a code-accessible representation of a convention we build here that says "timers all happen around here" and "network init all happens around here", which in turn is there to help us see more clearly what happens when, and also to allow applications to hook in code "before the first network is accessed".

The indirections I meant are going through Kconfig, then through definitions of initialization priorities, and then using them to define the entries.

>  There would probably no reason to reorder modules if dependencies were modeled correctly.

They are (hopefully) now, so that's a start for a sequence. Unlike on more complex systems with parallel initialization, there can be a constant total ordering of all startup components. (If any single autoinit component were so complex that it'd need to be placed at one point in one configuration and at another in another, I'd assume it can be split up into an early and a late phase.)

Reply:
Ah ok, groups of known points sounds good.